README.md
<!--
 @changed 2022.02.24, 04:22
-->

# Cam flask photo receiver server & raspberry camera client


## Build info (auto-generated)

- Version: 0.0.5
- Last changes timestamp: 2022.02.24, 00:55
- Last changes timetag: 220224-0055


## API

Basic api structure:

TODO: Describe api (for specific server version)


## Server urls

Remote server: `http://cam-rpi-server.lilliputten.ru/`


## Server

Images server runs on python/flask platform.

TODO: Describe basic server functionality.

TODO: Preserve capability to implement public server (with same or like functinal?)


## Python venv maintenance

Server command for creating venv:

```
virtualenv -p python3 ~/.venv-py3-flask
source ~/.venv-py3-flask/bin/activate
pip install -r requirements.txt
```

Local script for venv creating and initialization:

```
sh utils/util-venv-init.sh
```

Local command for activate venv:

```
call .venv/Scripts/activate
source .venv/Scripts/activate
```

## Python dependencies

```
pip install PKGNAME
pip freeze > requirements-frozen.txt
pip install -r requirements.txt
```

## Camera interface

Camera shots are taken using the `raspistill` program using commands like:

```shell
# Default:
raspistill -o image.jpg
# Half:
raspistill -w 1296 -h 972 -o image-half.jpg
# Quarter:
raspistill -w 648 -h 486 -o image-quarter.jpg
```

For commandline reference use `raspistill --help`.

- [Camera configuration - Raspberry Pi Documentation](https://www.raspberrypi.org/documentation/configuration/camera.md)


## Raspberry camera client

Use `client-make-and-upload-image.sh` script to make & upload image to server.

Use crontab to automate image capture.

### Sample crontab lines:

- Every minute: `* * * * * /home/pi/projects/cam-rpi-server/client-make-and-upload-image.sh`
- Every 15th minute: `*/15 * * * * /home/pi/projects/cam-rpi-server/client-make-and-upload-image.sh`

## Real crontab entry example:

```shell
# Test entry...
30 */1 * * * date >> ~/test_crontab
*/5 * * * * date >> ~/test_crontab

# Make & upload shots every 5 minutes (with forced logging)...
*/5 * * * * sh /home/pi/projects/cam-rpi-server/client-make-and-upload-image.sh >> /home/pi/projects/cam-rpi-server/cron-log.txt 2>&1

# Make & upload shots every 20 minutes...
*/20 * * * * sh /home/pi/projects/cam-rpi-server/client-make-and-upload-image.sh

# Reboot every 3 hours (00:55, 03:55, etc...)
55 */3 * * * sudo reboot -f

# Start server on boot
@reboot sh /home/pi/projects/cam-rpi-server/utils/rpi-start-server.sh
```

### Crontab commands:

Edit crontab:
```shell
crontab -e
```

Show crontab:
```shell
crontab -l
```

Automatically add server start command to crontab with:

```shell
sh utils/rpi-gunicorn-server-create-crontab-entry.sh
```

See also:

- [Scheduling tasks with Cron - Raspberry Pi Documentation](https://www.raspberrypi.org/documentation/linux/usage/cron.md)


## Crontab logging

Uncomment `# cron.*` line in `/etc/rsyslog.conf` (eg. edit with `sudo vim /etc/rsyslog.conf`).

Show crontab log:

```shell
tail -f /var/log/cron.log
```

Or use output reirect in command:

```shell
/home/pi/cam-rpi-server/client-make-and-upload-image.sh >> /home/pi/cam-rpi-server/cron.log 2>&1
python /home/pi/cam-client/client-make-image.py >>  /home/pi/projects/cam-rpi-server/cron.log 2>&1
```

## Run gunicorn service on raspberry (using `index.py`)

- [Beginners guide to install a Flask (Python) web server on Raspberry PI](https://peppe8o.com/beginners-guide-to-install-a-flask-python-web-server-on-raspberry-pi/)
- [Running Gunicorn — Gunicorn 20.1.0 documentation](https://docs.gunicorn.org/en/stable/run.html)

> gunicorn -w 1 -b 0.0.0.0:4000 hello:app

Real command (see `utils/rpi-start-server.sh`, attention for port number, in `-b` switch):

```
&& "$GUNICORN" \
  --reload \
  -D \
  -w 1 \
  -b 0.0.0.0:4000 \
  --chdir "$ROOT" \
  --pid="$PIDFILE" \
  --log-file="$LOGFILE" \
  index \
&& echo "Done." \
```

See `log-gunicorn.log` and `log-gunicorn.pid` files in project root folder after start.

For autostart see 'crontab' section above.


CHANGELOG.md
<!--
@changed 2022.02.27, 02:45
-->

# Changelog

- 2022.02.27, 02:45 -- RecordsStorage using TinyDB (fully worked code & tests).
- 2022.02.26, 05:31 -- Using TinyDB for data engine in `RecordsStorage` (in progress: basic methods and tests are ready).
- 2022.02.24, 00:55 -- v.0.0.5: Moved `lib` to `core/lib` scope (extract to shared submodule in future). Using pylint as project-wide linter (fixed all linter issues).
- 2022.02.23, 00:24 -- RecordsStorage: remove outdated records during findng (in `processRecords`), added metod for rorced remove of outdated records (`removeOutdatedRecords`).
- 2022.02.22, 02:30 -- RecordsStorage, Record: Basic records and records storage routines and tests.
- 2022.02.21, 23:06 -- Test support utilities (`getTrace`). Restuctured utils module.
- 2022.02.15, 06:12 -- v.0.0.4: Created automatic tests & linters for python code.
- 2022.02.15, 03:53 -- Helper modules moved to lib folder.
- 2022.02.14, 04:45 -- Experimantal sessions support.
- 2022.02.12, 04:52 -- Implemented simple socket.io tests. Relocated cdn files.
- 2022.02.12, 02:58 -- v.0.0.3: Restrucured server application (extracted several blueprint modules for different apis).
- 2022.02.08, 06:19 -- Device server shot creation method.
- 2022.02.08, 01:51 -- Fixed logger async (non-atomic) print issuesm server: starting only once (skip first initialization in debug mode, when app is initialized twice), templates: temporarily removed unused assets.
- 2022.02.07, 22:33 -- Local device server start scripts (via gunicorn, see `utils/rpi-start-server.sh`).
- 2022.02.07, 03:24 -- Readme venv instructions, disable basic auth in htaccess.


config.py
# -*- coding:utf-8 -*-
# @module config
# @desc Universal server & client config
# @since 2022.02.06, 23:56
# @changed 2022.02.26, 02:34
# See:
#  - https://docs.python.org/3/library/configparser.html -- ???
#  - https://stackoverflow.com/questions/9590382/forcing-python-json-module-to-work-with-ascii
#  - https://flask.palletsprojects.com/en/2.0.x/config/

import os
from os import path
import json
#  import yaml
import sys

from config_helpers import updateConfigWithYaml, readFiletoString

pythonVersion = sys.version

gunicornEnv = os.getenv('GUNICORN_ENV')
flaskEnv = os.getenv('FLASK_ENV')
isGunicorn = bool(gunicornEnv)
isRPi = isGunicorn
isDev = flaskEnv == 'development'
isProd = not isDev

rootPath = path.dirname(path.abspath(__file__))  # Project root path

yamlConfigFilename = path.join(rootPath, 'config.yml')
yamlLocalConfigFilename = path.join(rootPath, 'config.local.yml')

uploadPath = path.join(rootPath, 'uploads')

#  Build params...
clientTemplatePath = path.join(rootPath, 'cam-client-app-build')
clientStaticPath = path.join(rootPath, 'static')
#  clientStaticPath = path.join(clientTemplatePath, 'static')

#  Generate/read build parameters (version, timetag etc)
#  Default values (empty)...
version = ''
timestamp = ''
timetag = ''
buildTag = ''
#  Filenames...
buildVersionFilename = path.join(rootPath, 'build-version.txt')
buildTagFilename = path.join(rootPath, 'build-tag.txt')
timestampFilename = path.join(rootPath, 'build-timestamp.txt')
timetagFilename = path.join(rootPath, 'build-timetag.txt')
packageFilename = path.join(rootPath, 'package.json')
#  Read version...
#  print('config: packageFilename', packageFilename  # DEBUG)
if path.isfile(buildVersionFilename):
    version = readFiletoString(buildVersionFilename)
elif path.isfile(packageFilename):
    pkgConfigFile = open(packageFilename)
    pkgConfig = json.load(pkgConfigFile)
    version = pkgConfig['version'].encode('ascii')
    pkgConfigFile.close()
# Read timestamp/timetag...
if path.isfile(timestampFilename):
    timestamp = readFiletoString(timestampFilename)
if path.isfile(timetagFilename):
    timetag = readFiletoString(timetagFilename)
# Read/generate buildTag...
if version and timetag:
    buildTag = 'v.' + version + '-' + timetag
elif path.isfile(buildTagFilename):
    buildTag = readFiletoString(buildTagFilename)

config = {  # Default config

    # Application parameters...

    'flaskEnv': flaskEnv,
    'gunicornEnv': gunicornEnv,
    'isGunicorn': isGunicorn,
    'isRPi': isRPi,
    'isDev': isDev,
    'isProd': isProd,

    'pythonVersion': pythonVersion,
    'version': version,
    'timestamp': timestamp,
    'timetag': timetag,
    'buildTag': buildTag,

    # Path parameters...

    'rootPath': rootPath,  # RO!
    'uploadPath': uploadPath,

    # Generated client path (see `cam-client-app-build`, TODO?)

    'clientStaticPath': clientStaticPath,
    'clientTemplatePath': clientTemplatePath,

    # Image parameters...
    #
    #  - [Camera configuration - Raspberry Pi Documentation](https://www.raspberrypi.org/documentation/configuration/camera.md)
    #
    #  Default dimensions: 2592x1944
    #  Half size: 1296x972
    #  Quarter size: 648x486

    'imageWidth': 648,
    'imageHeight': 486,

    # Client parameters... (UNUSED)

    'localImageFile': 'local-image.jpg',
    'remoteUrl': 'https://cam.lilliputten.ru/upload',

    # Image file paramaters...

    'imageExt': '.image',
    'imagesIndex': 'index.txt',

    # Logging...

    'outputLog': True,  # Print log to stdout
    'outputColoredLog': True,  # Use rich output log format with `termcolor`
    'writeLog': True,  # Write log to external file
    'clearLogFile': True,  # Clear log file at start
    'logFileName': 'log.txt',  # Log file name (relative to `rootPath`!)

    # Datetime formats...

    'dateTagFormat': '%y%m%d-%H%M',  # eg: '220208-0204'
    'dateTagPreciseFormat': '%y%m%d-%H%M%S',  # eg: '220208-020423'
    'shortDateFormat': '%Y.%m.%d-%H:%M',  # eg: '2022.02.08-02:04'
    'preciseDateFormat': '%Y.%m.%d-%H:%M:%S',  # eg: '2022.02.08-02:04:23'
    'logDateFormat': '%y%m%d-%H%M%S-%f',  # eg: '220208-020423-255157'
    'detailedDateFormat': '%Y.%m.%d-%H:%M:%S.%f',  # eg: '2022.02.08-02:04:23.255157'

    # Databases...

    'dbPath': path.join(rootPath, 'db'),
    'dbExt': '.db',  # sqlite3 database file extensions

}


updateConfigWithYaml(config, yamlConfigFilename)
updateConfigWithYaml(config, yamlLocalConfigFilename)

#  #  DEBUG
#  print('Config:', config)
#  print('Done')
